### PR TITLE
DHP 284 Store tokens in s3

### PR DESCRIPTION
**Pull request by ncbui**
## Description of change
This PR makes it possible to store veteran's device tokens into an s3 bucket. 

## Things to know about this PR
Changes include:
* Added token storage service to manage writing to S3 if app deployed, else saves to local files (matching Spark Schema)
* Moved methods from `fitbit_service` into existing `fitbit_client`. 
* Removed `fitbit_service`and tests
* Added custom errors (MissingAuthError, TokenExchangeError, TokenStorageError)
* Refactored fitbit#callback flow to throw custom errors and escape the call chain when error occurs
* Added explicit error handling to existing methods
* Rspec tests for services and clients
* Added settings for the connected device s3 bucket has been added. 
* Existing feature flag:  dhp_connected_devices
* Existing sentry logging kept
* No Prometheus metrics collected nor Grafana dashboard added
* No Swagger docs that were updated? 
* No PII concerns or questions? No

### CODE REVIEW

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (reblg) [inline modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:30]: Woww! This function is so much easier to follow now

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (reblg) [inline modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:22]: does this still need to be implemented? or can the comment we removed / captured in a jira ticket?

File: modules/dhp_connected_devices/app/services/token_storage_service.rb
Comment (reblg) [inline modules/dhp_connected_devices/app/services/token_storage_service.rb:71]: should the `local` directory be added to `.gitignore`

File: modules/dhp_connected_devices/lib/fitbit/client.rb
Comment (reblg) [inline modules/dhp_connected_devices/lib/fitbit/client.rb:49]: Can this say: "Retrieves auth code from callback_params"

File: modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb
Comment (reblg) [inline modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:33]: does this mean "not running locally"?

Comment (reblg) [reply modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:33]: oh i see on 39 that's what it means

**Review by reblg — APPROVED** [review]
I left a few comments that are mostly nitpicks but otherwise this looks really good to me! nice job!

File: modules/dhp_connected_devices/lib/fitbit/client.rb
Comment (ncbui) [reply modules/dhp_connected_devices/lib/fitbit/client.rb:49]: Adding

File: modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb
Comment (ncbui) [reply modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:33]: yea, would adding 'not running locally' help our future selves too? I tried to avoid saying "developing in dev" bc VA.gov technically has a development env we could deploy too

Comment (reblg) [reply modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:33]: potentially but I think it's ok because you can easily get the context below

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb]: removing this note and adding to JIRA

**Review by yonidoronpeters — CHANGES_REQUESTED** [review]
Great job on this - very good refactor!
I still need to review the tests (will get to it later today). Please see comments. Most important one is the edge case not covered in the fitbit controller

File: modules/dhp_connected_devices/lib/fitbit/client.rb
Comment (ncbui) [reply modules/dhp_connected_devices/lib/fitbit/client.rb:13]: trying to make that note generic for both error types. Can update it the note to include both errors, but we could also remove that note

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:25]: Updating the note

Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:25]: @yonidoronpeters  how about "Fitbit callback error: #{error message}"

File: modules/dhp_connected_devices/app/services/token_storage_service.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/services/token_storage_service.rb:71]: added

File: modules/dhp_connected_devices/lib/fitbit/client.rb
Comment (ncbui) [reply modules/dhp_connected_devices/lib/fitbit/client.rb:32]: added

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (yonidoronpeters) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:25]: I like it!

File: modules/dhp_connected_devices/lib/fitbit/client.rb
Comment (ncbui) [reply modules/dhp_connected_devices/lib/fitbit/client.rb:13]: removed comment

File: modules/dhp_connected_devices/app/services/token_storage_service.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/services/token_storage_service.rb]: Removed this step after updating adding 'symbolize_names: true' to JSON.parse() command in get_tokens

File: modules/dhp_connected_devices/app/services/token_storage_service.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/services/token_storage_service.rb]: Updated this to `return store_locally(current_user, device_key, payload_json) unless vsp_env_exists`

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb]: Added a `unless storage_status == false` here after updating token service to return immediately if storing locally.

Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb]: Could we revisit this discussion with the disconnect device story? Can add a note on that ticket

File: modules/dhp_connected_devices/app/services/token_storage_service.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/app/services/token_storage_service.rb:8]: This is not needed anymore as far as I can tell. We only log to Sentry in the controller

File: modules/dhp_connected_devices/lib/fitbit/client.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/lib/fitbit/client.rb:28]: mentioned it previously, but probably worth including the resp body as well here for more detail about what happened

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:21]: If storage `status == false`, we redirect with status error which is great. But we don't actually know that this happened because we never log anything for this one case. Trying to think of a clean way to also log something to the extent of `"Token storage failed for user with ICN: #{@current_user&.icn}"` so that at least we know someone tried to connect their Fitbit and it failed on the token persistence step

File: modules/dhp_connected_devices/spec/controllers/dhp_connected_devices/fitbit/fitbit_controller_spec.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/spec/controllers/dhp_connected_devices/fitbit/fitbit_controller_spec.rb:85]: Nice! I really like this improvement

File: modules/dhp_connected_devices/spec/controllers/dhp_connected_devices/fitbit/fitbit_controller_spec.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/spec/controllers/dhp_connected_devices/fitbit/fitbit_controller_spec.rb:149]: Nitpick comment: this SentryLogging mocking is not needed in this case.

File: modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb:87]: NIT: can we give a different value? e.g. 'code_value' or '12345'. Might make this test a bit clearer

File: modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb:97]: Question out of curiosity : is there a reason we us `do ... end` block instead of `{}` for the last case?

File: modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:47]:
```suggestion
      it 'returns true when the token was stored locally' do
```

File: modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb
Comment (yonidoronpeters) [inline modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:40]: It is worth testing the sad path (error thrown) in local storage

**Review by yonidoronpeters — APPROVED** [review]
Added a few more comments, but they are small and should be quick to address. Approving so that you can merge right after addressing those.
Excellent job on this story! 💯 🎉

File: modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb
Comment (ncbui) [reply modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb:87]: Definitely, updating

File: modules/dhp_connected_devices/spec/controllers/dhp_connected_devices/fitbit/fitbit_controller_spec.rb
Comment (ncbui) [reply modules/dhp_connected_devices/spec/controllers/dhp_connected_devices/fitbit/fitbit_controller_spec.rb:149]: Removed

File: modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb
Comment (ncbui) [reply modules/dhp_connected_devices/spec/lib/fitbit_client_spec.rb:97]: This was a rubocop suggestion to work with line length. I've parameterized the error again, so the style can go back to the `{}`

File: modules/dhp_connected_devices/app/services/token_storage_service.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/services/token_storage_service.rb:8]: removing

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:21]: I'm open to it. Splitting this off as a tiny PR for the story.

File: modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb
Comment (ncbui) [reply modules/dhp_connected_devices/spec/services/dhp_connected_devices/token_storage_service_spec.rb:40]: Added

File: modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb
Comment (ncbui) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:21]: Narrator voice: we did not split this off. Raising and error instead of returning false

Comment (yonidoronpeters) [reply modules/dhp_connected_devices/app/controllers/dhp_connected_devices/fitbit/fitbit_controller.rb:21]: Great, looks good. In this case, we can replace this line with 
```suggestion
redirect_with_status('success')
```
and we have one let scenario to test 😄

**Pull request merged**
